compiler/rustc_mir/src/transform/early_otherwise_branch.rs

```diff
@@ -270,6 +260,16 @@ impl<'a, 'tcx> Helper<'a, 'tcx> {
                 return None;
             }
 
+            // The types of the two adts matched on have to be equal for this optimization to apply
+            if discr_info.type_adt_matched_on != this_bb_discr_info.type_adt_matched_on {
+                trace!(
+                    "NO: types do not match. LHS: {:?}, RHS: {:?}",
+                    discr_info.type_adt_matched_on,
+                    this_bb_discr_info.type_adt_matched_on
+                );
+                return None;
+            }
```

Comment:
There doesn't seem to be much point to this change without benchmarks. Personally I would expect a simple `!=` to be faster than two conditionals.